README.md
# CommonCodes #

[version_shield]: https://img.shields.io/badge/version-2.2.0-blue.svg
[latest_release]: https://mfederczuk.github.io/commoncodes "Latest Release"
[![version: 2.1.0][version_shield]][latest_release]
[![Changelog](https://img.shields.io/badge/-Changelog-blue.svg)](./CHANGELOG.md "Changelog")

## About ##

**CommonCodes** is an attempt to create a standardized list of program exit
 statuses.

---

So one day I wondered if there was any standard for exit statuses/codes, because
 I, like (hopefully) many other people like following pre-existing standards.  
I was rather disappointed to see that there wasn't really anything like that.  
I could only find **Bash**s definition for codes above 125, that were used for
 special cases like termination through a sent signal, and
 [`sysexits.h`](https://man.openbsd.org/sysexits), which is kinda what I was
 looking for, but there weren't many definitions, and the ones that did exist
 were just too generic.

Now, you can't really call this a proper standardization. I wanted a full list
 of statuses, where every code from 0 to 255 was defined to have some meaning.
 But this was just something that did not exist... until now.

## Contributing ##

Read through the [CommonCodes Contribution Guidelines](./CONTRIBUTING.md)
 if you want to contribute to this project.

## License ##

[GNU GPLv3+](./LICENSE)


CHANGELOG.md
<!-- markdownlint-disable MD024 -->

# Changelog #

All notable changes to this project will be documented in this file.
The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [2.2.0] - 2021-11-13 ##

[2.2.0]: https://github.com/mfederczuk/commoncodes/compare/v2.1.0...v2.2.0

### Changed ###

* status 9 (`may not be empty`) message
  * `[<option>: ]argument [<n>: ]may not be (empty|blank)` → `[<option>: ]argument [<n>: ](may|must) not be (empty|blank)`

## [2.1.0] - 2021-11-01 ##

[2.1.0]: https://github.com/mfederczuk/commoncodes/compare/v2.0.0...v2.1.0

### Changed ###

* status 101 (`not enough memory`) message and description
  * message: `not enough [heap ]memory` → `not enough [(heap|kernel) ]memory`
  * description: _"**[...]** free memory on the heap to allocate a block of data."_ →
     _"**[...]** free memory (either heap or kernel memory) to perform an operation."_

## [2.0.0] - 2020-04-13 ##

[2.0.0]: https://github.com/mfederczuk/commoncodes/compare/v1.0.0...v2.0.0

### Added ###

* new status 98: `<path>: too many levels of symbolic link`
* new status 99: `<filename>: filename too long`
* new status 26: `<item>: <itemtype> already exists`
* new status 27: `<program>: program missing`
* `glob(7)` and `regex(7)` to **See Also** page section
* to status range 32 - 47 (`(custom feedback statuses)`) description
  * _"**[...]** These should always be documented."_
* to status 30 (`connection timed out`) message and description
  * message: **`[...]`**`[: waited <ms> ms]`
  * description: _"**[...]**¶&lt;ms&gt; may be how many milliseconds were waited."_

### Changed ###

* message format explanation in description
* statuses 26 (`network error[: <info>]`) - 29 (`arithmetic error[: <info>]`)
  positions to 28 - 31
* status range 79 - 98 (`(custom configuration errors)`) to 79 - 97 (-1)
* statuses 99 (`memory error`) - 102 (`generic internal fault`) positions to
  100 - 103
* status range 103 - 122 (`(custom internal faults)`) to 104 - 122 (-1)
* swap status 25 (`item not an itemtype`) and status 26 (`item already exists`)
* status 103 (`generic internal fault`) description
  * _"General internal fault.**[...]**"_ → _"Generic internal fault.**[...]**"_
* status 9 (`may not be empty`) description
  * _"Program can not **[...]**"_ → _"The program can not **[...]**"_
* status 6 (`unexpected option`) description
  * _"The seen option at this position with the set of combination and **[...]**"_ →
     _"The spotted option at this position with the combination and **[...]**"_

### Fixed ###

* wording in description
* status 11 (`out of range`) description (typo fix)
  * _"**[...]** be the minumum, floor, maximum **[...]**"_ →
     _"**[...]** be the minimum, floor, maximum **[...]**"_

### Removed ###

* status 30: `divided by 0 error`
* status 31: `(overflow|underflow) error`

## [1.2.0] - 2021-11-13 ##

[1.2.0]: https://github.com/mfederczuk/commoncodes/compare/v1.1.0...v1.2.0

### Changed ###

* status 9 (`may not be empty`) message
  * `[<option>: ]argument [<n>: ]may not be (empty|blank)` → `[<option>: ]argument [<n>: ](may|must) not be (empty|blank)`

## [1.1.0] - 2021-11-01 ##

[1.1.0]: https://github.com/mfederczuk/commoncodes/compare/v1.0.0...v1.1.0

### Changed ###

* status 100 (`not enough memory`) message and description
  * message: `not enough [heap ]memory` → `not enough [(heap|kernel) ]memory`
  * description: _"**[...]** free memory on the heap to allocate a block of data."_ →
     _"**[...]** free memory (either heap or kernel memory) to perform an operation."_

## [1.0.0] - 2019-05-11 ##

[1.0.0]: https://mfederczuk.github.io/commoncodes/v/0.3.0...1.0.0.html

### Added ###

* new status 6: `<option>: unexpected option`
* new status 8: `<argument>: unknown [sub]command`
* new status 11: `[<option>: ]<argument>: out of range[ (<relational_op> <limit>)]`
* new statuses 32 - 47: `(custom feedback error)`
* to status 3 (`missing arguments`) description
  * _"**[...]**¶&lt;arguments&gt; may be the placeholder name of the missing arguments."_
* to status 4 (`too many arguments`) description
  * _"**[...]**¶&lt;n&gt; may be the amount of arguments that are not needed."_
* to status 7 (`may not be empty`) description
  * _"**[...]**¶&lt;n&gt; is the position of the argument that is empty/blank and is only really needed when the program/option needs multiple arguments."_
* `[: <info>]` and `<option>` placeholder explanation to description
* `grep(1)` to "**See Also**" page section

### Changed ###

* status 6 (`invalid argument`) position to 7
* statuses 7 (`may not be empty`) & 8 (`not a number`) positions to 9 & 10
* status 9 (`does not match`) position to 12
* status range 10 - 15 (`(custom usage errors)`) to 13 - 23 (+5)
* statuses 16 (`no such itemtype`) & 17 (`not an itemtype`) positions to 24 & 25
* statuses 22 (`network error`) - 27 (`overflow/underflow error`) positions to
  26 - 31
* statuses 28 (`memory error`) - 30 (`stack overflow`) positions to 99 - 101
* status 31 (`emergency stop`) position to 123
* status range 32 - 63 (`(custom errors)`) to 48 - 63 (-16)
* status range 79 - 99 (`(custom configuration errors)`) to 79 - 98 (-1)
* status 100 (`generic internal fault`) position to 102
* status range 101 - 123 (`(custom internal faults)`) to 103 - 122 (-2)
* status 3 (`missing arguments`) message
  * `[<option>: ]missing argument(s)[: <arguments>...]` →
    `[<option>: ]missing argument[s][: <argument>...]`
* status 7 (`may not be empty`) message
  * `argument #<n>: may not be (empty|blank)` →
    `[<option>:]argument [<n>:]may not be (empty|blank)`

### Removed ###

* from status 8 (`not a number`) description
  * _"**[...]**¶If a valid decimal number has been entered and this exit status is returned then you must convert the decimal number into an integer."_

## [0.3.0] - 2019-05-06 ##


commoncodes.h
/*
 * CommonCodes C header file - Standardized list of commonly used exit statuses.
 * Copyright (C) 2019 Michael Federczuk
 * 
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#ifndef _COMMONCODES_H
#define _COMMONCODES_H

/* See <https://speziil.github.io/commoncodes/v/1.0.0.html> for more info */

#define CC_SUCCESS        0 /* success */
#define CC_GEN_ERR        1 /* generic error */
#define CC_GEN_USAGE_ERR  2 /* generic usage error */
#define CC_MISS_ARGS      3 /* missing arguments */
#define CC_TOO_MANY_ARGS  4 /* too many arguments */
#define CC_INV_OPT        5 /* invalid option */
#define CC_UNEX_OPT       6 /* unexpected option */
#define CC_INV_ARG        7 /* invalid argument */
#define CC_UNKNOWN_SUBCMD 8 /* unknown subcommand */
#define CC_UNKNOWN_CMD    8 /* unknown command */
#define CC_NO_EMPTY_ARG   9 /* may not be empty */
#define CC_NO_BLANK_ARG   9 /* may not be blank */
#define CC_NO_NUM         10 /* not a number */
#define CC_NO_INT         10 /* not an integer */
#define CC_OUT_OF_RANGE   11 /* out of range */
#define CC_NO_MATCH       12 /* does not match */
/* 13 - 23: custom usage errors */
#define CC_NO_SUCH_ITEMTYPE 24 /* no such itemtype */
#define CC_NOT_A_ITEMTYPE   25 /* not a itemtype */
#define CC_NOT_AN_ITEMTYPE  25 /* not an itemtype */
#define CC_NET_ERR          26 /* network error */
#define CC_NO_NET_CON       27 /* no network connection */
#define CC_CON_TIMED_OUT    28 /* connection timed out */
#define CC_ARIT_ERR         29 /* arithmetic error */
#define CC_DIV_BY_0_ERR     30 /* divied by 0 error */
#define CC_OFLOW_ERR        31 /* overflow error */
#define CC_UFLOW_ERR        31 /* underflow error */
/* 32 - 47: custom feedback statuses */
/* 48 - 63: custom errors */
/* BSD errors */
#define CC_USAGE       64 /* command line usage error */
#define CC_DATAERR     65 /* data format error */
#define CC_NOINPUT     66 /* cannot open input */
#define CC_NOUSER      67 /* addresse unknown */
#define CC_NOHOST      68 /* host name unknown */
#define CC_UNAVAILABLE 69 /* service unavailable */
#define CC_SOFTWARE    70 /* internal software error */
#define CC_OSERR       71 /* system error (e.g., can't fork) */
#define CC_OSFILE      72 /* critical OS file missing */
#define CC_CANTCREAT   73 /* can't create (user) output file */
#define CC_IOERR       74 /* inout/output error */
#define CC_TEMPFAIL    75 /* temp failure; user is invited to retry */
#define CC_PROTOCOL    76 /* remote error in protocol */
#define CC_NOPERM      77 /* permission denied */
#define CC_CONFIG      78 /* configuration error */
/* BSD errors end */
/* 79 - 98: custom configuration errors */
#define CC_MEM_ERR           99 /* memory error */
#define CC_NO_MEM           100 /* not enough memory */
#define CC_NO_HMEM          100 /* not enough heap memory */
#define CC_STACK_OFLOW_ERR  101 /* stack overflow error */
#define CC_GEN_INTERN_FAULT 100 /* generic internal fault */
/* 101 - 122: custom internal faults */
#define CC_EMERG_STOP                      123 /* emergency stop */
#define CC_SCRIPT_CALLED_INTERACTIVELY     124 /* script was called interactively */
#define CC_SCRIPT_NOT_CALLED_INTERACTIVELY 124 /* script was not called interactively */
#define CC_UNKNOWN_ERR                     125 /* unknown error */

#endif /* _COMMONCODES_H */
